shift_detector/precalculations/TextMetadata.py

```diff
@@ -0,0 +1,383 @@
+import unicodedata
+from abc import abstractmethod
+from collections import defaultdict
+
+import pandas as pd
+# noinspection PyPackageRequirements
+from iso639 import languages
+from langdetect import detect, DetectorFactory
+from nltk.corpus import stopwords
+# noinspection PyPackageRequirements
+from spellchecker import SpellChecker
+from textstat import textstat
+
+from shift_detector.precalculations.Precalculation import Precalculation
+from shift_detector.utils import UCBlist
+from shift_detector.utils.ColumnManagement import ColumnType
+from shift_detector.utils.TextMetadataUtils import dictionary_to_sorted_string, tokenize, delimiters
+
+
+class GenericTextMetadata(Precalculation):
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__)
+
+    def __hash__(self):
+        return hash(self.__class__)
+
+    @staticmethod
+    @abstractmethod
+    def metadata_name() -> str:
+        pass
+
+    @abstractmethod
+    def metadata_column_type(self) -> ColumnType:
+        pass
+
+    @abstractmethod
+    def metadata_function(self, text):
+        pass
+
+    def process(self, store):
+        metadata1 = pd.DataFrame()
+        metadata2 = pd.DataFrame()
+        df1, df2 = store[ColumnType.text]
+        for column in df1.columns:
+            clean1 = df1[column].dropna()
+            clean2 = df2[column].dropna()
+            logger.info(self.metadata_name() + ' analysis for ' + column)
+            metadata1[column] = [self.metadata_function(text) for text in clean1]
+            metadata2[column] = [self.metadata_function(text) for text in clean2]
+        return metadata1, metadata2
+
+
+class NumCharsMetadata(GenericTextMetadata):
+
+    @staticmethod
+    def metadata_name() -> str:
+        return 'num_chars'
+
+    def metadata_column_type(self) -> ColumnType:
+        return ColumnType.numerical
+
+    def metadata_function(self, text):
+        return len(text)
+
+
+class RatioUpperMetadata(GenericTextMetadata):
+
+    @staticmethod
+    def metadata_name() -> str:
+        return 'ratio_upper'
+
+    def metadata_column_type(self) -> ColumnType:
+        return ColumnType.numerical
+
+    def metadata_function(self, text):
+        if text == "":
+            return 0
+        lower = sum(map(str.islower, text))
+        upper = sum(1 for c in text if c.isupper())
+        return upper / (lower + upper)
+
+
+class UnicodeCategoriesMetadata(GenericTextMetadata):
+
+    @staticmethod
+    def metadata_name() -> str:
+        return 'unicode_categories'
+
+    def metadata_column_type(self) -> ColumnType:
+        return ColumnType.categorical
+
+    @staticmethod
+    def unicode_category_histogram(text):
+        characters = defaultdict(int)
+        for c in text:
+            category = unicodedata.category(c)
+            characters[category] += 1
+        return characters
+
+    def metadata_function(self, text):
+        return dictionary_to_sorted_string(self.unicode_category_histogram(text))
```

Comment:
@velramiir does some statistical tests with this and by converting the dictionary to a string, he can use the result as a categorical value instead of using every value of the dictionary as a numeric value. The String only contains the keys of the dictionary, sorted by their value, so they are comparable.